Use optimized overflow operation everywhere

celinval opened the pull request:
### Description of changes: 

This is related to the performance degradation we are seeing in the toolchain upgrade: https://github.com/model-checking/kani/pull/2293

### Resolved issues:

Towards https://github.com/model-checking/kani/issues/1483. There are a few remaining instances of add_overflow and mul_overflow in intrinsics.rs.

### Related RFC:

<!--
Link to the Tracking RFC issue if this work implements part of an RFC.
-->
Optional #ISSUE-NUMBER.

### Call-outs:

This PR has a side effect of removing redundant checks inserted by CBMC for checked arithmetic operations.

### Testing:

* How is this change tested? Existing tests and manual tests with my other branch where I'm doing the upgrade

* Is this a refactor change? yes

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval commented:
It looks like the performance is not always better... :(

celinval commented:
I do wonder if this proof is also somewhat unstable. I noticed that in my other PR (#2407) that had minimum changes to the model, there was also a performance degradation of 47% for the same test.

tautschnig commented on cprover_bindings/src/goto_program/expr.rs:
```diff
@@ -1448,6 +1456,7 @@ impl Expr {
 
     /// `ArithmeticOverflowResult r; >>>r.overflowed = builtin_sub_overflow(self, e, &r.result)<<<`
     pub fn mul_overflow(self, e: Expr) -> ArithmeticOverflowResult {
+        // TODO: Should we always use the one below?
         let result = self.clone().mul(e.clone());
         let overflowed = self.mul_overflow_p(e);
         ArithmeticOverflowResult { result, overflowed }
```
I'd answer that "TODO" with a "yes, unless there really is a need to only compute the Boolean would-this-operation-overflow result." That's not the case here, which does the multiplication, so that's likely just a left-over from the time where we didn't yet have the OverflowResult* expressions. We should get rid of this.

zhassan-aws replied:
Yes, the intention was to replace all the usage of `mul_overflow` with `mul_overflow_result` and get rid of `mul_overflow`.

tautschnig commented on tests/ui/cbmc_checks/signed-overflow/expected:
```diff
@@ -7,12 +7,5 @@ Failed Checks: attempt to calculate the remainder with a divisor of zero
 Failed Checks: attempt to calculate the remainder with overflow
 Failed Checks: attempt to shift left with overflow
 Failed Checks: attempt to shift right with overflow
-Failed Checks: arithmetic overflow on signed addition
-Failed Checks: arithmetic overflow on signed subtraction
-Failed Checks: arithmetic overflow on signed multiplication
-Failed Checks: shift distance is negative
-Failed Checks: shift distance too large
-Failed Checks: shift operand is negative
 Failed Checks: arithmetic overflow on signed shl
-Failed Checks: shift distance is negative
-Failed Checks: shift distance too large
```
Isn't that absence of failing checks concerning?

zhassan-aws replied:
I think those checks were redundant, i.e. each failure condition was being checked twice.

celinval replied:
Sorry, I should've added that to the call-out. Yes, @zhassan-aws is correct that these checks were redundant.

These operations that we changed are checked operations, which means that the codegen adds an overflow check already. Since the implementation does not invoke the arithmetic operation directly, CBMC doesn't add its own overflow check anymore.

That said, I think I deleted some of the shift errors by accident. They shouldn't be affected.

tautschnig replied:
Ok, thanks, I should have looked at more context in that file (which has all those codegen-added overflow checks also as "Failed Checks".

tautschnig reviewed (changes requested):
I'd like to ask for an explanation as to why the no-longer-failing checks are ok.

zhassan-aws commented on cprover_bindings/src/goto_program/expr.rs:
```diff
@@ -1403,9 +1403,9 @@ impl Expr {
     }
 
-    /// Uses CBMC's add-with-overflow operation that performs a single addition
+    /// Uses CBMC's [binop]-with-overflow operation that performs a single addition
```
```suggestion
    /// Uses CBMC's [binop]-with-overflow operation that performs a single arithmetic
```

zhassan-aws reviewed (approved):
This has been on my to-do list for too long. Thanks for taking care of it @celinval!

You should add in the PR description "Towards https://github.com/model-checking/kani/issues/1483". There are a few remaining instances of `add_overflow` and `mul_overflow` in `intrinsics.rs` (and possibly elsewhere).

celinval commented:
> I'd like to ask for an explanation as to why the no-longer-failing checks are ok.

That's fair! My bad, I should've added an explanation before. I added it as part of an answer to one of your question and addressed the issue in the call-out section. In a nutshell, the arithmetic check that was being added by CBMC was redundant.
